Install service dependencies

kvikshaug opened the pull request:
* All service dependencies are now installed in the base image
* Pip cache is kept in the image to speed up `pip-compile` step in child images
* Upgraded alpine 3.8 -> 3.10

kvikshaug committed:
chore: update apk package db

kvikshaug committed:
chore: move workdir command

kvikshaug committed:
chore: rename wsgi-requirements -> base-requirements

kvikshaug committed:
docs: comment on build commands

kvikshaug committed:
chore: upgrade to alpine 3.10

kvikshaug committed:
chore: check gunicorn version, not help output

This doesn't have any functional effect, but it's a bit more succinct.

kvikshaug committed:
feat: install all expected service dependencies

kvikshaug committed:
chore: don't remove pip cache

kvikshaug committed:
chore: add compatibility constraints

kvikshaug committed:
docs: add ADR

Midnighter commented on alpine/Dockerfile:
```diff
@@ -12,23 +12,46 @@
+    ca-certificates \
+    # Libraries required for psycopg2
+    postgresql-dev \
+    python3-dev \
```
I think the headers should be available without the system package. The Python docker images in general install their own version into the base system and do not use system packages.

kvikshaug replied:
:+1: I'll remove it. Not a fan of circumventing the distro's packaging system though.

kvikshaug replied:
I'm getting the same error as below when `python3-dev` is not installed:

    ImportError: Error loading shared library libpq.so.5: No such file or directory (needed by /usr/local/lib/python3.6/site-packages/psycopg2/_psycopg.cpython-36m-x86_64-linux-gnu.so)

Do you have any advice as for how to use the preinstalled headers? I really feel the best and least time-consuming approach here is to just install and keep the packages.

Midnighter replied:
I'll give this a test. I guess we would need `libpq-dev` at compile time and `libpq` only at runtime. I guess the headers wouldn't be too big so it might be easiest to keep them around.

Midnighter replied:
Okay, I think I got it sorted out but will comment on the postgres-base PR.

Midnighter commented on alpine/Dockerfile:
```diff
@@ -12,23 +12,46 @@
+    # Libraries required for psycopg2
+    postgresql-dev \
+    python3-dev \
+    musl-dev
```
If we want to keep these, it seems to me that they should be added to the `.buiild-deps` later on so that they will be removed again.

kvikshaug replied:
Are you sure that even `postgresql-dev` isn't also needed at runtime?

kvikshaug replied:
I'm getting this error at runtime, so I think the library needs to stay installed.

    ImportError: Error loading shared library libpq.so.5: No such file or directory (needed by /usr/local/lib/python3.6/site-packages/psycopg2/_psycopg.cpython-36m-x86_64-linux-gnu.so)

Midnighter replied:
See above.

Midnighter commented on alpine/Dockerfile:
```diff
@@ -12,23 +12,46 @@
 WORKDIR /opt
 
-COPY wsgi-requirements.in ./
+RUN apk update
```
I'm not sure about the interaction of this and the `--no-cache` flag later on. Would the flag force another update step each?

Midnighter replied:
Also, keeping the list files around is not necessary which is why one typically rolls update, install, and clean up into one Docker layer.

kvikshaug replied:
Ah, indeed `--no-cache` forces updates every step, so that's a very unhelpful addition. I'll remove it.

Note that the list files consume 1.42MB, so it's a bit of a bikeshedding issue.

Midnighter reviewed (changes requested):
1. I don't think it's the best idea to have the base image include all the postgres dependencies, too, since we do have microservices that do not use postgres. So maybe the `postgres-base` image could build on top of this image instead?
2. I'm also not sure about all the development dependencies. Isn't that something that should be handled in the cookiecutter. They really don't need to be part of production images in my opinion.
3. While I understand the appeal of keeping the pip cache around, I wonder if that actually works the way you anticipate. What if the cache from the base image is a week or a month old, would it still lead to the same improvements? I guess it might speed up hashing the dependencies.

I guess a big question for me is where do you want to see the speed ups? Is it for local development or do you want to increase throughput for CI&CD?

Lastly, I propose the following enhancements:

All our services contain the following in their Dockerfiles:

```dockerfile
ENV APP_USER=giraffe

ARG UID=1000
ARG GID=1000

ARG CWD=/app
ENV PYTHONPATH=${CWD}/src
WORKDIR "${CWD}"

RUN addgroup -g "${GID}" -S "${APP_USER}" && \
    adduser -u "${UID}" -G "${APP_USER}" -S "${APP_USER}"
```

this can become part of the base image.

I would also propose to add the following statement to the Dockerfile or it should maybe better go into the cookiecutter Dockerfile.

```dockerfile
EXPOSE 8000
```

kvikshaug commented:
> 1. I don't think it's the best idea to have the base image include all the postgres dependencies, too, since we do have microservices that do not use postgres. So maybe the `postgres-base` image could build on top of this image instead?

Yes! Sorry, I forgot about that image. I don't think anyone made an effort to apply it in any services yet, but it makes sense to apply it now.

> 2. I'm also not sure about all the development dependencies. Isn't that something that should be handled in the cookiecutter. They really don't need to be part of production images in my opinion.

They already are; this PR only addresses the issue of moving the installation to the base image.

> 3. While I understand the appeal of keeping the pip cache around, I wonder if that actually works the way you anticipate. What if the cache from the base image is a week or a month old, would it still lead to the same improvements? I guess it might speed up hashing the dependencies.

* The `pip-compile` step in the services is constrained by the exact versions installed in the base image and so the cache will always be used regardless of age.
* During image building (`pip-sync`), if a new version for some dependency was released, *and* updated in the base image, but we haven't recompiled requirements for the service, then the cache will not be used, and `pip-sync` will uninstall the newer version and replace it with the older version. This gives us an incentive to keep our packages up-to-date, something we could later try to automate if we have enough confidence in our QA and test suite.

> I guess a big question for me is where do you want to see the speed ups? Is it for local development or do you want to increase throughput for CI&CD?

We sacrifice both image size and build time of the wsgi-base image (since those are run daily by cron), in order to speed up builds of all service images, both locally and on CI.

> Lastly, I propose the following enhancements:

Good ideas. The scope of this PR is limited to python package dependency handling though so feel free to raise these proposals as issues or PRs.

kvikshaug committed:
chore: remove empty continuation lines

kvikshaug committed:
feat: don't include DB dependencies

kvikshaug committed:
chore: remove `apk update` step

Midnighter commented:
> They already are; this PR only addresses the issue of moving the installation to the base image.

Right, not completely happy about that but I can see how it's a productivity improvement.

Midnighter commented on docs/architecture/0003-install-all-base-dependencies.md:
```diff
@@ -0,0 +1,27 @@
+
+* WSGI libraries
+* Flask and related plugins
+* Database packages
```
```suggestion
```

Midnighter commented on alpine/base-requirements.in:
```diff
@@ -0,0 +1,35 @@
+# flake8-docstrings fixes compatibility with v4.0.0
+pydocstyle==3.0.0
+
+# apispec and marshmallow (dependencies of flask-apispec): Upgrading to latest
```
Kinda sad that flask-apispec doesn't pin them itself :/

kvikshaug replied:
Well, we haven't really investigated why the latest versions break. Could very well be a backwards incompatibility that we just need to address in our code. But yes if they're simply incompatible, that would be sad!

Midnighter commented on alpine/base-requirements.in:
```diff
@@ -0,0 +1,35 @@
+# Compatibility constraints
+
+# pydocstyle (dependency of flake8-docstrings): pin to v3.0.0 until
+# flake8-docstrings fixes compatibility with v4.0.0
```
I'm not sure this is still an issue.

kvikshaug replied:
I can try it out separately after all this is merged.

Midnighter reviewed (approved):
Cheers, I'm glad we agree on most points. Made a few more comments but handle them as you see fit.

kvikshaug committed:
feat: add debian image

- Changes image tag from branch-based to `alpine`
- Adds a `debian` tag for the debian version

kvikshaug committed:
docs: db packages are not installed now

Co-Authored-By: Moritz E. Beber <morbeb@biosustain.dtu.dk>

kvikshaug committed:
chore: tag images with commit hash
